src/metabase/pulse/render/body.clj

```diff
@@ -1030,16 +1030,31 @@
       [:img {:style (style/style {:display :block :width :100%})
              :src   (:image-src image-bundle)}]]}))
 
+(defn- all-unique?
+  [funnel-rows]
+  (let [ks (into #{} (map :key funnel-rows))]
+    (= (count ks) (count funnel-rows))))
+
+(defn- build-funnel-rows
+  [funnel-rows raw-rows]
+  (let [funnel-rows (map-indexed (fn [idx row] (assoc row :idx (inc idx))) funnel-rows)
+        xf-rows     (into {} (map (fn [row] (vec (take-last 2 row))) raw-rows))]
+    (for [{idx     :idx
+           k       :key
+           enabled :enabled} funnel-rows]
+      (when enabled
+        [k (or (get xf-rows k) (get xf-rows idx))]))))
```

Comment:
This fallback feels a bit magical to me, especially since I don't imagine we'd want to support a mixture of keys and indexes in our raw rows.

Just food for thought - did you consider the approach where you pick a case at the start, based on whether the raw rows have any `string?` values?